From 55e297cc379321e93f93de1a7618117442459f7a Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Fri, 13 Oct 2023 14:09:59 +0200
Subject: [PATCH 417/841] feat(stm32mp2): display pending IAC in BL2 entry

Dump the IAC in BL2 entry to detect issues in low power entry sequence,
including the IAC generated by the ROM code.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 621a58f243aca33d2d0d4aedde708c9f95685f58 feat(stm32mp2): display pending IAC in BL2 entry]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I23c5fbf4bf2e27361b5346c5b80aff5cdd606525
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/331718
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/stm32mp2/bl2_plat_setup.c | 34 +++++++++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index fb35d9a2b..d93e3a199 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -49,6 +49,38 @@ IMPORT_SYM(uintptr_t, __BSS_END__, BSS_END);
 IMPORT_SYM(uintptr_t, __DATA_START__, DATA_START);
 IMPORT_SYM(uintptr_t, __DATA_END__, DATA_END);
 
+#define IAC_EXCEPT_LSB_BIT(x) ((x) * 32U)
+#define IAC_EXCEPT_MSB_BIT(x) (IAC_EXCEPT_LSB_BIT(x) + 31U)
+
+static void iac_dump(void)
+{
+#if DEBUG
+	unsigned int i;
+	unsigned int bit;
+
+	for (i = 0U; i < IAC_NB; i++) {
+		uint32_t isr = mmio_read_32(IAC_BASE + IAC_ISR(i));
+		uint32_t ier = mmio_read_32(IAC_BASE + IAC_IER(i));
+
+		isr &= ier;
+		if (isr == 0U) {
+			continue;
+		}
+
+		WARN("IAC exceptions pending [%u:%u] = %x\n",
+		     IAC_EXCEPT_MSB_BIT(i), IAC_EXCEPT_LSB_BIT(i), isr);
+		bit = 0U;
+		while ((isr != 0U) && (bit < 32U)) {
+			if ((isr & BIT(0)) != 0U) {
+				WARN(" - %u\n", IAC_EXCEPT_LSB_BIT(i) + bit);
+			}
+			isr >>= 1U;
+			bit++;
+		}
+	}
+#endif
+}
+
 static void print_reset_reason(void)
 {
 	uint32_t rstsr = mmio_read_32(stm32mp_rcc_base() + RCC_C1BOOTRSTSCLRR);
@@ -266,6 +298,8 @@ void bl2_el3_plat_arch_setup(void)
 		goto skip_console_init;
 	}
 
+	iac_dump();
+
 	stm32mp_print_cpuinfo();
 
 	board_model = dt_get_board_model();
-- 
2.47.2

